RFC: config: Make vfio_mode=vfio by default

dgibson opened the pull request:
Kata currently defaults to `vfio_mode=guest-kernel`.  This mode has its uses but is unlike (and can never be like) the behaviour with a normal runC container.  By contrast `vfio_mode=vfio` enables handling of VFIO devices in a way that resembles runC in at least some cases (and more cases might be added in future by removing Kata limitations).

Therefore we should default to vfio_mode=vfio instead.  To make that work, we also need to enable a vIOMMU as well, so we need to update the qemu configuration file for that as well.  Note that enabling a vIOMMU in the guest doesn't compel the guest kernel to use it for all devices, so this should not adversely impact performance.

NOTE: This should work well for qemu.  Firecracker and ACRN don't support VFIO devices in any form at present, so the change is irrelevant for them.  That leaves cloud-hypervisor; it should theoretically allow `vfio_mode=vfio`, but will require an vIOMMU to do so.  I'm not familiar enough with clh to know exactly what needs to be done here - googling suggests that clh supports virtio-iommu, but I'm not clear if that's true in the version we currently use, or if it needs some configuration change to enable it.

dgibson commented:
/test

dgibson commented:
/retest-arm
/retest-s390x
/retest-metrics

dgibson commented:
/retest-ubuntu-metrics

sboeuf commented:
Hi @dgibson, can you please clarify a few things:
- What does `vfio` mode mean? Is it meant for exposing a device behind an IOMMU inside the VM so that VFIO can be used from the VM directly?
-  And on the other hand, does `guest-kernel` mean a VFIO device is passed through so that it won't be attached to any IOMMU from a guest perspective?

I'm concerned that you want to enable a vIOMMU by default since this obviously means any device placed behind this vIOMMU will have degraded performance compared to a device that would not be placed behind a vIOMMU.
If we take the example of a NIC, the buffers can be allocated anywhere in guest RAM, meaning every time this device is used, the vIOMMU will have to update the mapping in the DMAR of the physical IOMMU. This takes time and is the main reason why you will never be able to achieve native performance with that device.
From a user perspective, the point is to run workloads as fast as possible, therefore I don't think they would choose to use a VFIO device directly from within the container. Instead, it's better if they can rely on the NIC or GPU that has been passed through based on the VFIO framework.
Enabling a vIOMMU should be used in certain niche use cases, but I don't think it should be the default.

One alternative to expose a VFIO device inside the VM without taking the hit on performance would be to rely on VFIO_NOIOMMU in the guest. Since we would have already passed the device through the VM, we would know it's entirely assigned to the VM and totally safe thanks to the physical IOMMU. Therefore, it won't be a security issue to bind this same device to VFIO with the NOIOMMU option in the guest.

From a CH perspective, we do support vIOMMU through `virtio-iommu`.

dgibson commented:
> Hi @dgibson, can you please clarify a few things:
> 
>     * What does `vfio` mode mean? Is it meant for exposing a device behind an IOMMU inside the VM so that VFIO can be used from the VM directly?

Semantically, it means that if a vfio device appears in the container spec, that vfio device appears within the container as `/dev/vfio/XXX` just as it would with runC (well, almost).

In implementation that means we need to rebind the device to `vfio_pci` driver within the guest, which in turn requires a vIOMMU.

>     * And on the other hand, does `guest-kernel` mean a VFIO device is passed through so that it won't be attached to any IOMMU from a guest perspective?

`guest-kernel` has no precise semantics.  The device is passed to the VM, and the guest kernel interprets it as it wills.  Typically that will mean some kind of device or network interface will become available to the container, but exactly what depends on many details.  A vIOMMU is not typically required, but that's irrelevant to the semantics of the device.

As described in #4332, this is inherently Kata specific behaviour, which can never conform to any semblance of OCI semantics.

> I'm concerned that you want to enable a vIOMMU by default since this obviously means any device placed behind this vIOMMU will have degraded performance compared to a device that would not be placed behind a vIOMMU. 

At least with the typical Intel/x86 iommu model, there isn't really a virtual hardware level concept of "behind" an iommu.  The vIOMMU is available to the guest, but it's up to the guest kernel whether it places a given device into an iommu domain, or leaves it without iommu translation. AIUI, with normal configuration, Linux will leave the vIOMMU bypassed for most devices, but will use it for vfio devices.

> If we take the example of a NIC, the buffers can be allocated anywhere in guest RAM, meaning every time this device is used, the vIOMMU will have to update the mapping in the DMAR of the physical IOMMU. This takes time and is the main reason why you will never be able to achieve native performance with that device. 

Well, maybe.  Even when using the vIOMMU, Linux will typically establish mappings for all of guest memory quite early, avoiding frequent updates.

For `vfio_mode=vfio` that's under control of the workload/application.  AIUI DPDK, at least, will also establish mappings for its whole buffer pool early, again avoiding frequent updates.

> From a user perspective, the point is to run workloads as fast as possible, therefore I don't think they would choose to use a VFIO device directly from within the container. Instead, it's better if they can rely on the NIC or GPU that has been passed through based on the VFIO framework. 
 
As described in #4332, using `guest-kernel` requires a Kata-aware workload which specifically knows how to locate and use whatever device the guest kernel presents.

Workloads written for vfio devices in a normal (runC) environment will expect to see vfio devices within the container, which will only appear in Kata with `vfio_mode=vfio`.  Regardless of iommu setup, this can allow better performance because it allows the workload to access hardware essentially directly, bypassing both guest and host kernels.

> Enabling a vIOMMU should be used in certain niche use cases, but I don't think it should be the default.

At least with qemu "enabling" the iommu here just means making it available to the guest, it doesn't force it to use it.  I'm unclear on how this works with clh, but afaik virtio-iommu is still up to the guest to choose to use for specific devices.

> One alternative to expose a VFIO device inside the VM without taking the hit on performance would be to rely on VFIO_NOIOMMU in the guest. Since we would have already passed the device through the VM, we would know it's entirely assigned to the VM and totally safe thanks to the physical IOMMU. Therefore, it won't be a security issue to bind this same device to VFIO with the NOIOMMU option in the guest.

Uh.. I believe noimmu mode requires the workload/application to specifically invoke it, so that won't work for existing workloads made to work with runC.

> From a CH perspective, we do support vIOMMU through `virtio-iommu`.

Does that require any startup options to enable?  Or is it always available to guests? Can the guest use virtio-iommu with any device, or only with specifically selected devices?

sboeuf commented:
> > Hi @dgibson, can you please clarify a few things:
> > ```
> > * What does `vfio` mode mean? Is it meant for exposing a device behind an IOMMU inside the VM so that VFIO can be used from the VM directly?
> > ```
> 
> Semantically, it means that if a vfio device appears in the container spec, that vfio device appears within the container as `/dev/vfio/XXX` just as it would with runC (well, almost).
> 
> In implementation that means we need to rebind the device to `vfio_pci` driver within the guest, which in turn requires a vIOMMU.
> 
> > ```
> > * And on the other hand, does `guest-kernel` mean a VFIO device is passed through so that it won't be attached to any IOMMU from a guest perspective?
> > ```
> 
> `guest-kernel` has no precise semantics. The device is passed to the VM, and the guest kernel interprets it as it wills. Typically that will mean some kind of device or network interface will become available to the container, but exactly what depends on many details. A vIOMMU is not typically required, but that's irrelevant to the semantics of the device.
> 
> As described in #4332, this is inherently Kata specific behaviour, which can never conform to any semblance of OCI semantics.
> 

Ok thanks for the clarification, I understand now why you're trying to enable this behavior by default.

> > I'm concerned that you want to enable a vIOMMU by default since this obviously means any device placed behind this vIOMMU will have degraded performance compared to a device that would not be placed behind a vIOMMU.
> 
> At least with the typical Intel/x86 iommu model, there isn't really a virtual hardware level concept of "behind" an iommu. The vIOMMU is available to the guest, but it's up to the guest kernel whether it places a given device into an iommu domain, or leaves it without iommu translation. AIUI, with normal configuration, Linux will leave the vIOMMU bypassed for most devices, but will use it for vfio devices.
> 

Sure but in the case you describe, you're expecting the device not to be bypassed since you need it to be placed behind the vIOMMU, otherwise you coudn't bind it to VFIO.

> > If we take the example of a NIC, the buffers can be allocated anywhere in guest RAM, meaning every time this device is used, the vIOMMU will have to update the mapping in the DMAR of the physical IOMMU. This takes time and is the main reason why you will never be able to achieve native performance with that device.
> 
> Well, maybe. Even when using the vIOMMU, Linux will typically establish mappings for all of guest memory quite early, avoiding frequent updates.
> 

This is highly dependent on the type of device, and if you think about production use cases, the most common device that's being passed through is a NIC, and unfortunately in that case Linux can't establish the mappings in advance.

> For `vfio_mode=vfio` that's under control of the workload/application. AIUI DPDK, at least, will also establish mappings for its whole buffer pool early, again avoiding frequent updates.
> 

Yes I agree DPDK is a good example of how the networking buffer limitation can be worked around.

> > From a user perspective, the point is to run workloads as fast as possible, therefore I don't think they would choose to use a VFIO device directly from within the container. Instead, it's better if they can rely on the NIC or GPU that has been passed through based on the VFIO framework.
> 
> As described in #4332, using `guest-kernel` requires a Kata-aware workload which specifically knows how to locate and use whatever device the guest kernel presents.
> 
> Workloads written for vfio devices in a normal (runC) environment will expect to see vfio devices within the container, which will only appear in Kata with `vfio_mode=vfio`. Regardless of iommu setup, this can allow better performance because it allows the workload to access hardware essentially directly, bypassing both guest and host kernels.
> 
> > Enabling a vIOMMU should be used in certain niche use cases, but I don't think it should be the default.
> 
> At least with qemu "enabling" the iommu here just means making it available to the guest, it doesn't force it to use it. I'm unclear on how this works with clh, but afaik virtio-iommu is still up to the guest to choose to use for specific devices.
> 
> > One alternative to expose a VFIO device inside the VM without taking the hit on performance would be to rely on VFIO_NOIOMMU in the guest. Since we would have already passed the device through the VM, we would know it's entirely assigned to the VM and totally safe thanks to the physical IOMMU. Therefore, it won't be a security issue to bind this same device to VFIO with the NOIOMMU option in the guest.
> 
> Uh.. I believe noimmu mode requires the workload/application to specifically invoke it, so that won't work for existing workloads made to work with runC.
> 

Yes I get it, this won't fit the "standard" runC use case.

> > From a CH perspective, we do support vIOMMU through `virtio-iommu`.
> 
> Does that require any startup options to enable? Or is it always available to guests? Can the guest use virtio-iommu with any device, or only with specifically selected devices?

It's similar to what you described with Intel vIOMMU, as it depends if the guest has the virtio-iommu driver or not, and if it does, it might decide to attach the device to an IOMMU domain or not. To have this working without interfering with other devices which we definitely don't want to see attached behing a vIOMMU, we need the user to create a dedicated PCI segment for the devices which need to be placed behind the vIOMMU.

I believe someone needs to update the `clh.go` code in order to make sure these two `vfio_mode` modes are handled correctly.

dgibson commented:
> > > I'm concerned that you want to enable a vIOMMU by default since this obviously means any device placed behind this vIOMMU will have degraded performance compared to a device that would not be placed behind a vIOMMU.
> > 
> > At least with the typical Intel/x86 iommu model, there isn't really a virtual hardware level concept of "behind" an iommu. The vIOMMU is available to the guest, but it's up to the guest kernel whether it places a given device into an iommu domain, or leaves it without iommu translation. AIUI, with normal configuration, Linux will leave the vIOMMU bypassed for most devices, but will use it for vfio devices.
> 
> Sure but in the case you describe, you're expecting the device not to be bypassed since you need it to be placed behind the vIOMMU, otherwise you coudn't bind it to VFIO.

The kernel can rearrange IOMMU domains at runtime, so typically it will disable IOMMU bypass on the device only when it is bound to the VFIO driver.  The point is, we don't need to decide which devices will and won't use the IOMMU at VM construction time

> > > If we take the example of a NIC, the buffers can be allocated anywhere in guest RAM, meaning every time this device is used, the vIOMMU will have to update the mapping in the DMAR of the physical IOMMU. This takes time and is the main reason why you will never be able to achieve native performance with that device.
> > 
> > 
> > Well, maybe. Even when using the vIOMMU, Linux will typically establish mappings for all of guest memory quite early, avoiding frequent updates.
> 
> This is highly dependent on the type of device, and if you think about production use cases, the most common device that's being passed through is a NIC, and unfortunately in that case Linux can't establish the mappings in advance.

Uh.. maybe not in all cases, but I'm pretty sure it can in most (with modern 64-bit hardware, at any rate).  But note that even if there's a cost to enabling the vIOMMU for a NIC, this will only affect the case where we're using the weird `guest-kernel` semantics.  That requires a Kata-aware workload and configuration anyway, so adding an explicit disable of the IOMMU via annotation doesn't seem much of an extra ask.

> > For `vfio_mode=vfio` that's under control of the workload/application. AIUI DPDK, at least, will also establish mappings for its whole buffer pool early, again avoiding frequent updates.
> 
> Yes I agree DPDK is a good example of how the networking buffer limitation can be worked around.
> 
> > > From a user perspective, the point is to run workloads as fast as possible, therefore I don't think they would choose to use a VFIO device directly from within the container. Instead, it's better if they can rely on the NIC or GPU that has been passed through based on the VFIO framework.
> > 
> > 
> > As described in #4332, using `guest-kernel` requires a Kata-aware workload which specifically knows how to locate and use whatever device the guest kernel presents.
> > Workloads written for vfio devices in a normal (runC) environment will expect to see vfio devices within the container, which will only appear in Kata with `vfio_mode=vfio`. Regardless of iommu setup, this can allow better performance because it allows the workload to access hardware essentially directly, bypassing both guest and host kernels.
> > > Enabling a vIOMMU should be used in certain niche use cases, but I don't think it should be the default.
> > 
> > 
> > At least with qemu "enabling" the iommu here just means making it available to the guest, it doesn't force it to use it. I'm unclear on how this works with clh, but afaik virtio-iommu is still up to the guest to choose to use for specific devices.
> > > One alternative to expose a VFIO device inside the VM without taking the hit on performance would be to rely on VFIO_NOIOMMU in the guest. Since we would have already passed the device through the VM, we would know it's entirely assigned to the VM and totally safe thanks to the physical IOMMU. Therefore, it won't be a security issue to bind this same device to VFIO with the NOIOMMU option in the guest.
> > 
> > 
> > Uh.. I believe noimmu mode requires the workload/application to specifically invoke it, so that won't work for existing workloads made to work with runC.
> 
> Yes I get it, this won't fit the "standard" runC use case.

Right.  I feel like we should match expected container semantics first, optimize performance second.

> > > From a CH perspective, we do support vIOMMU through `virtio-iommu`.
> > 
> > 
> > Does that require any startup options to enable? Or is it always available to guests? Can the guest use virtio-iommu with any device, or only with specifically selected devices?
> 
> It's similar to what you described with Intel vIOMMU, as it depends if the guest has the virtio-iommu driver or not, 

We control the guest kernel, so we can make sure that's the case.

> and if it does, it might decide to attach the device to an IOMMU domain or not. To have this working without interfering with other devices which we definitely don't want to see attached behing a vIOMMU, we need the user to create a dedicated PCI segment for the devices which need to be placed behind the vIOMMU.

So.. I don't really see why a special PCI segment is needed.  Surely virtio-iommu, like the Intel IOMMU, allows domains to be constructed of (mostly) arbitrary sets of devices.  We control the guest kernel, and we already rely on it for many things, I don't see why we shouldn't trust it to make sensible performance choices about which devices to use a vIOMMU for.

Also.. having multiple PCI segments on clh is a problem.  Something I noticed in earlier reworks is that the clh PCI hotplug API appears to be broken such that it can't safely be used to plug in anything on a non-root bus.  AFAICT we only get away with it because currently everything *is* on the root bus.

> I believe someone needs to update the `clh.go` code in order to make sure these two `vfio_mode` modes are handled correctly.

Mostly that should be done.  When I reworked the PCI hotplug handling I adjusted the clh code as well, though I didn't and don't have the ability to test beyond what the CI covers.  The actual `vfio_mode=vfio` support isn't really hypervisor specific, it's about reconfiguring the devices once they appear to the guest kernel (it's almost entirely in the agent not the runtime).

The sole remaining question for clh, AFAICT, is how to make sure an IOMMU is available to the guest.

dgibson commented:
/retest-fc

dgibson commented:
/retest-s390x

dgibson commented:
/retest-ubuntu-metrics

zvonkok commented:
Several questions: 
> behaviour with a normal runC container

If a VFIO device is mounted into the container this would mean this very VFIO device would appear in the container, something like this? 
```
$ sudo ctr  run   --device /dev/vfio/71 --rm -t  "docker.io/library/archlinux:latest" arch ls /dev/vfio/71 
/dev/vfio/71
```
What would be the use-case to have the VFIO device inside a Kata container? Nested virtualization? 

For all things related to GPUs and other passthrough devices how should this then work? I do not need the VFIO device, I want the actual device. 

A normal runC container is not consuming VFIO devices you would mount `/dev/nvidia` and not `/dev/vfio/71`. 

On use-case I know of where the "raw" VFIO device is needed in a container is KubeVirt, but this VFIO is not used by the container directly it is passed to the VM running inside the Pod. 

I hope that we're still leaving `guest-kernel` as a valid option. Removing that would mean that we would need different container images when running natively versus with Kata.  That is the beauty of the current implementation I can take any preexisting GPU container image and run it without modifications on Kata.

zvonkok commented:
> Because it relies on guest kernel interpretation of the device it doesn't - and never can - match the behaviour of a traditional (runC or crun) container

I do not see why it has to, we should look at the problem from the container aka workload point of view, and not how something is realized. As said before currently we can run HW enabled containers without modification and I think this has the biggest value. It would compare runC or crun with the kata-agent and in this regard kata-agent is doing exactly was runC or crun are doing. 

> Whether a device appears in the container, what its name and type will be and in some cases even the device interface will vary depending on the guest kernel version and configuration

Can you give an example here? 

> Because it has inherently Kata specific behaviour, general Kubernetes plugins will not generally be able to provide meaningful device location information to workloads, and in some cases may apply actively misleading information

How will passing a VFIO into the container make this better? Which plugins are you talking about? Device-Plugin? CPU Manager? Topology Manager?  Isn't that all related to the VMM and how it setups the guest topology? 

> The upshot of this is that to effectively use vfio_mode=guest-kernel requires a tailored, kata-aware workload, and carefully controlled Kata configuration (including the guest kernel image in use).

Exactly the opposite I can run my CUDA/DOCA/RAPIDS application containers 1:1 with Kata. 

> By contrast, vfio_mode=vfio is - at least in theory - allows Kata to be a drop in replacement for runC containers using vfio devices (such as those managed by the k8s sr-iov device plugin). Several limitations prevent that working in all cases at present, but it will work in simple cases, and many of those limitations can lifted with Kata improvements.

This would be one example where Kata would need to provide the raw VFIO devices, e.g. when used with Multus and this would need the IOMMU and then create the interfaces that the Pod would get attached. 

Are you saying you want to run the sriov-device-plugin as a Kata container? Who is going to consume the VFs? Multus?

zvonkok commented on src/runtime/Makefile:
```diff
@@ -197,7 +197,7 @@ DEFVALIDVHOSTUSERSTOREPATHS := [\"$(DEFVHOSTUSERSTOREPATH)\"]
 DEFVALIDFILEMEMBACKENDS := [\"$(DEFFILEMEMBACKEND)\"]
 DEFMSIZE9P := 8192
-DEFVFIOMODE := guest-kernel
+DEFVFIOMODE := vfio
```
What implications does this have to dynamically switch from one to another? Can I still use a separate config file to enable guest-kernel?

dgibson replied:
Yes, you can still change the config file to use `guest-kernel`, this is just the default.  In addition you can override this on a per-pod basis using the `io.katacontainers.config.runtime.vfio_mode` annotation.

dgibson commented:
> Several questions:
> 
> > behaviour with a normal runC container
> 
> If a VFIO device is mounted into the container this would mean this very VFIO device would appear in the container, something like this?
> 
> ```
> $ sudo ctr  run   --device /dev/vfio/71 --rm -t  "docker.io/library/archlinux:latest" arch ls /dev/vfio/71 
> /dev/vfio/71
> ```

Yes, just like that.

> What would be the use-case to have the VFIO device inside a Kata container? Nested virtualization?

The main one we have in mind is containerized network functions based on DPDK.  By passing in a VFIO device, DPDK's userspace drivers can more-or-less directly access the hardware for high performance.  There are probably other applications using vfio-based userspace drivers, though I don't know what they are off hand.

Nester virtualization could cover quite a few different cases, some of which also might have use for passing VFIO devices into a container like this.

> For all things related to GPUs and other passthrough devices how should this then work? I do not need the VFIO device, I want the actual device.

I'm not at all familiar with the GPU use cases, though I understand that it is a main user of the current `guest-kernel` behaviour, and will probably want to continue using that mode at least for now.

> A normal runC container is not consuming VFIO devices you would mount `/dev/nvidia` and not `/dev/vfio/71`.

Right... and AIUI you would specify `/dev/nvidia` as the device to be passed into the container, not a VFIO device.  Currently, I believe that won't work with Kata, for now, anyway.  Basically each device type that could be passed into a Kata container needs specialized handling and `/dev/nvidia` is not one of them, although I think it could be added (possibly using VFIO to pass the device to the VM internally).

> On use-case I know of where the "raw" VFIO device is needed in a container is KubeVirt, but this VFIO is not used by the container directly it is passed to the VM running inside the Pod.

That's another possible use case yes.  The VM is part of the workload, so the container *is* using the VFIO device.

> I hope that we're still leaving `guest-kernel` as a valid option. Removing that would mean that we would need different container images when running natively versus with Kata. That is the beauty of the current implementation I can take any preexisting GPU container image and run it without modifications on Kata.

I'm only suggesting changing a default here, not removing the `guest-kernel` option.  Note that this can already be changed on a per-pod basis using the `io.katacontainers.config.runtime.vfio_mode` annotation.

dgibson commented:
> > Because it relies on guest kernel interpretation of the device it doesn't - and never can - match the behaviour of a traditional (runC or crun) container
> 
> I do not see why it has to do that in the first place, we should look at the problem from the container aka workload point of view, and not how something is realized. 

Yes... well, I think we should consider both the workload point of view and the point of view of the person or tools configuring the pods.  That's exactly why I'm focusing on container semantics, not performance or VM specific considerations.

> As said before currently we can run HW enabled containers without modification and I think this has the biggest value. It would compare runC or crun with the kata-agent and in this regard kata-agent is doing exactly was runC or crun are doing.

So, as I said, I'm not familiar with the GPU based use cases.  The only way I can see that this would work in `guest-kernel` mode with unmodified workloads is if for the runC case you're passing in `/dev/nvidia` but for the Kata case you/re passing in `/dev/vfio/whatever`.  So while the workloads are the same, the pod configuration is on.  Even then, I think this is only going to work because there's a single well known device name - anything where the specific device is parameterized isn't going to work, because in `guest-kernel` mode Kata has no control (or even visibility) of what the device nodes will be, so it has no way to make them match those given in the container spec.

`guest-kernel` absolutely does not match runC container if you do pass in `/dev/vfio/whatever`.  For runC, you get `/dev/vfio/whatever` in the container, just like any other device.  With `guest-kernel` you get... something.. depending on what device is passed in and the guest kernel configuration.

> > Whether a device appears in the container, what its name and type will be and in some cases even the device interface will vary depending on the guest kernel version and configuration
> 
> Can you give an example here?

The most obvious case is if the guest kernel just doesn't have a driver for the device, in which case nothing at all will appear in the container.  A subtler case is that sometimes the default naming of devices has changed with kernel versions.  In that case the name of the device in the container will also vary depending on the guest kernel version, and Kata has no visibility or control over that.

A different device interface is rarer, I'll admit, but there are a few cases of hw devices which have multiple different drivers presenting different sw device interfaces to userspace.

> > Because it has inherently Kata specific behaviour, general Kubernetes plugins will not generally be able to provide meaningful device location information to workloads, and in some cases may apply actively misleading information
> 
> How will passing a VFIO into the container make this better?

This is, alas, imperfect even with `vfio` mode.  The SR-IOV device plugin provides `PCIDEVICE_xxx` environment variables with PCI addresses of the devices.  The `vfio_mode=vfio` code looks for such environment variables and translates them to guest PCI addresses.  We could do the same if a plugin provided a `/dev/vfio/whatever` path, though we don't do so at present.

> Which plugins are you talking about? Device-Plugin? CPU Manager? Topology Manager? Isn't that all related to the VMM and how it setups the guest topology?

Chiefly the SR-IOV device plugin.  See for example [here](https://developers.redhat.com/articles/2021/08/27/using-virtual-functions-dpdk-red-hat-openshift) . That's with runC and openshift, but with `vfio_mode=vfio` it should be possible to drop in Kata without workload or pod configuration changes.

> > The upshot of this is that to effectively use vfio_mode=guest-kernel requires a tailored, kata-aware workload, and carefully controlled Kata configuration (including the guest kernel image in use).
> 
> Exactly the opposite I can run my CUDA/DOCA/RAPIDS application containers 1:1 with Kata.

Interesting.  I think this is relying on specific behaviours of the CUDA device, and will not generalize.  Even then, as above, I'm pretty sure this must require a pod configuration that's different between runC and Kata.

> > By contrast, vfio_mode=vfio is - at least in theory - allows Kata to be a drop in replacement for runC containers using vfio devices (such as those managed by the k8s sr-iov device plugin). Several limitations prevent that working in all cases at present, but it will work in simple cases, and many of those limitations can lifted with Kata improvements.
> 
> Are you saying you want to run the sriov-device-plugin as a Kata container? Who is going to consume the VFs? Multus?

No.  The sriov-device-plugin (in `vfio-pci` mode, rather than `netdevice` mode) simply provides VFs in exactly the same way as it does for runC and Kata is able to consume them.

zvonkok commented:
Ok, this makes all sense, I just wanted to emphasize the usage of both modes and even in DPDK, you can run workloads with or without IOMMU, meaning you would need both the `vfio` and `guest-kernel` mode. 

Often times DPKD is run with an IOMMU to have better protection of devices aka isolated DMAs. Either you want isolation or performance, or a thing in between with `iommu=pt` where DMAs are just passthrough'ed and not translated. 

DPDK is also using the IOMMU to have in some cases contiguous memory but it is not a  hard requirement in the latest DPDK release this requirement was loosened. 

Anyway, we need both modes depending on how the devices are consumed, adding the IOMMU optionally will cover more use cases than we have now.

dgibson commented:
> Ok, this makes all sense, I just wanted to emphasize the usage of both modes and even in DPDK, you can run workloads with or without IOMMU, meaning you would need both the `vfio` and `guest-kernel` mode.

The modes aren't really about the IOMMU.  `vfio` does require a guest IOMMU, but the important difference is the semantics of how the device will appear in the container.

> Often times DPKD is run with an IOMMU to have better protection of devices aka isolated DMAs. Either you want isolation or performance, or a thing in between with `iommu=pt` where DMAs are just passthrough'ed and not translated.

If DPDK is using the VFIO driver backend then you need an IOMMU (well.. unless you use noiommu, but that needs a specially built workload).  DPDK does have other driver backends, but at least ones which use a netdevice are likely to hit bigger performance issues than the IOMMU with Kata, since they won't be acceessing the hardware directly.
 
> DPDK is also using the IOMMU to have in some cases contiguous memory but it is not a hard requirement in the latest DPDK release this requirement was loosened.
> 
> Anyway, we need both modes depending on how the devices are consumed, adding the IOMMU optionally will cover more use cases than we have now.

zvonkok commented:
If you do a passthrough of an MLX in guest-kernel mode you would have the full performance of the `netdevice`. As you said if you use the `VFIO` backend in DPDK then you need the IOMMU as mentioned before, we need both depending on how the application needs or wants to see a specific device. 

We're on the same page but just looking from different angles :)

dgibson commented:
> If you do a passthrough of an MLX in guest-kernel mode you would have the full performance of the `netdevice`.

Oof... so there are a lot of layers of complexity in that seemingly straightforward assertion.

- AIUI, Mellanox VFs generally do use netdevices with DPDK.  I'm not quite sure how that works to be performant, but it doesn't matter for these purposes.  Because a netdevice is being passed into the container, `vfio_mode` is irrelevant.  You pass a net device in, you get a net device in the container.
- There are a couple of different ways Kata can implement this.  The safe way is to bridge the host netdevice to a `virtio-net` in the VM, but that probably doesn't perform great.  In `physical_endpoint.go` there is logic to instead locate the undelying PCI device, pass it to the VM with VFIO and represent as a netdevice in the container.  Now, from some reports I've heard, I suspect that logic isn't working, and it's *definitely* conceptually broken (having ownership of the netdevice does not imply ownership of the PCI device).  Regardless, `vfio_mode` is *still* irrelevant; it only affects the handling of VFIO devices passed in as VFIO devices (that is `/dev/vfio/whatever` in the `ociSpec.Linux.Devices[]` list.
- Intel VFs with DPDK *do* use `vfio-pci` mode typically, with the container workload itself expecting a `/dev/vfio/whatever`.  This will only work in Kata with `vfio_mode=vfio` which requires an IOMMU.  For Kata we always require a host IOMMU anyway, but with this case we will also require a guest vIOMMU.  The cost of that should be far less than the cost of inserting the guest (and maybe the host's) net device layer into the stack.
- All of this is hard to get a handle on because the various layers of k8s, CNI and whatever device plugins are involved makes it non-trivial to figure out exactly what is being sent in the runtime container spec itself (whether that is runC or Kata).

> As you said if you use the `VFIO` backend in DPDK then you need the IOMMU as mentioned before, we need both depending on how the application needs or wants to see a specific device. DPDK is just one example.

In the short to medium term, that's true - `guest-kernel` has real use cases now.  However those *do* require either a different workload image and/or a different pod configuration to runC.  `guest-kernel` semantics just don't match anything runC can possibly do, since it involves one type of (sw) device in the container as viewed by the layers above, but a different type of (sw) device in the container as viewed by the workload.

The fact that it works with minimal changes for the GPU case is, AFAICT, basically a lucky accident.  I think the correct way to handle that long term would be to have Kata specifically handle `/dev/nvidia` (or whatever other devices) and pass them through to the VM and container as needed.  That *might* involve VFIO internally, but would not be affected by `vfio_mode`, which, again *only* affects the case of `/dev/vfio/whatever` being explicitly listed in the container configuration.

> We're on the same page but just looking from different angles :)

dgibson commented:
/test

dgibson commented:
/retest-arm

dgibson commented:
/retest-ubuntu-metrics

bergwolf commented on src/runtime/config/configuration-qemu.toml.in:
```diff
@@ -246,7 +246,7 @@ vhost_user_store_path = "@DEFVHOSTUSERSTOREPATH@"
 # This will also add the following options to the kernel's
 # command line: intel_iommu=on,iommu=pt
-#enable_iommu = true
+enable_iommu = true
```
There should be a check to assert that `enable_iommu` and `vfio_mode` are set as expected.

dgibson replied:
I'm not exactly sure what you mean by this.  Do you mean give an error if `enable_iommu = false` and `vfio_mode = vfio` at the same time?  The difficulty with this is that `enable_iommu` is qemu specific, whereas `vfio_mode` is generic.

bergwolf replied:
Yes, that's what I mean. Just wanted to rule out the possibility of mis-config rather than having to find it out upon container app failures. I think it can be done in `LoadConfiguration`, no?

bergwolf reviewed (commented):
`defaultVfioMode` in `pkg/katautils/config-settings.go.in` also needs to be changed. It's a bit sad that we have two places to define a single default value.

dgibson commented:
/test

fidencio commented:
@dgibson, we're closing this PR as it's been opened and inactive for more than one year.
If you think this is a mistake and the PR should be re-opened, please, feel free to do so after rebasing it atop of current main.

dgibson commented:
@fidencio no worries.  At this point I really don't care one way or another.
